### PR TITLE
GT_Recipe Speedup

### PR DESCRIPTION
**Pull request by TimeConqueror**

### CODE REVIEW

File: src/main/java/gregtech/api/util/GT_Utility.java [hunk 1]
```diff
@@ -1236,13 +1238,19 @@ public static boolean listContains(Object aObject, Collection aObjects) {
         return false;
     }
 
+    /**
+     * Note: use {@link ArrayExt#withoutNulls(Object[], IntFunction)} if you want an array as a result.
```
Comment (Glease) [inline src/main/java/gregtech/api/util/GT_Utility.java:1242]: FYI I did see this before and tried to optimize this. IIRC my tests showed the performance difference is minimal in most typical use case, i.e. arrays with 0~8 elements and occasionally 1 or 2 nulls. In certain scenario my "imporved" implementation even performed worse.

Of course our exact implementation differs a bit (I was using java 8 streams so do have some overhead), so just FYI.

File: src/main/java/gregtech/api/util/extensions/IteratorExt.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package gregtech.api.util.extensions;
+
+import gregtech.api.objects.iterators.ArrayIterator;
+import gregtech.api.objects.iterators.MergedIterator;
+import gregtech.api.objects.iterators.NonNullIterator;
+
+import java.util.Iterator;
+
+public class IteratorExt {
```
Comment (Glease) [inline src/main/java/gregtech/api/util/extensions/IteratorExt.java:9]: Didn't see this class used anywhere. Did you forget to push some files?

Comment (TimeConqueror) [reply src/main/java/gregtech/api/util/extensions/IteratorExt.java:9]: I wanted to remove some other stuff, but decided to not do so for now. But these classes can still be helpful.

File: src/main/java/gregtech/api/util/GT_Utility.java (hunk 1 above)
Comment (TimeConqueror) [reply src/main/java/gregtech/api/util/GT_Utility.java:1242]: By using #withoutNulls you creates only one new array, while when you use old method you create more:
1) new ArrayList<>(Collection<? extends E> c) calls c.toArray(), which copies the array.
2) ArrayList#remove also copies some part of array.

File: src/main/java/gregtech/api/util/extensions/IteratorExt.java (hunk 2 above)
Comment (Glease) [reply src/main/java/gregtech/api/util/extensions/IteratorExt.java:9]: On second thought these are either present in guava or base java library.

```java
// merge(itera, iterb, ...) equivalent
com.google.common.collect.Iterators.concat(itera, iterb, ...);
// withoutNulls(iter) equivalent
com.google.common.collect.Iterators.filter(iter, Objects::nonNull);
// ofArray(a, b, ...) equivalent
Arrays.asList(a, b, ...).iterator()
```

They might be *slightly* less performant, with all the built in null chceks and whatnot, but I don't think that would matter on anything better than celeron.

Comment (TimeConqueror) [reply src/main/java/gregtech/api/util/extensions/IteratorExt.java:9]: Let's check:
1) MergeIterator:
```
public static <T> Iterator<T> concat(Iterator<? extends T>... inputs) {
    return concat(ImmutableList.copyOf(inputs).iterator());
  }
```

```
public static <E> ImmutableList<E> copyOf(E[] elements) {
    switch (elements.length) {
      case 0:
        return ImmutableList.of();
      case 1:
        return new SingletonImmutableList<E>(elements[0]);
      default:
        return new RegularImmutableList<E>(checkElementsNotNull(elements.clone()));
    }
  }
```
As you can see in the last method `elements.clone()` is called, which copies the full array using the reflection as well.

2) Didn't know about #filter. I checked its implementation, it's fine. NonNullIterator is not needed. 

3) ofArray - yep, you're right, for some reason I didn't check `asList(a, b, ...).iterator`'s implementation, it's almost the same. NonNullIterator is not needed.

**Review by Glease — APPROVED** [review]

**Pull request merged**